x/rollapp/keeper/state_info.go

```diff
@@ -85,3 +93,50 @@ func (k Keeper) GetAllStateInfo(ctx sdk.Context) (list []types.StateInfo) {
 
 	return
 }
+
+// DeleteStateInfoUntilTimestamp deletes all stateInfo until the given timestamp
+func (k Keeper) DeleteStateInfoUntilTimestamp(ctx sdk.Context, endTimestampExcl time.Time) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.StateInfoKeyPrefix))
+	storeTS := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.TimestampedStateInfoKeyPrefix))
+
+	// get all latest stateInfo and latest finalized stateInfo and add them to a map
+	latestIndexes := append(k.GetAllLatestStateInfoIndex(ctx), k.GetAllLatestFinalizedStateIndex(ctx)...)
+	skipStateInfoIndexesMap := make(map[string]uint64, len(latestIndexes))
+	for _, index := range latestIndexes {
+		skipStateInfoIndexesMap[index.RollappId] = index.Index
+	}
+
+	k.IterateStateInfoWithTimestamp(storeTS, endTimestampExcl.UnixMicro(), func(keyTS []byte) bool {
+		key := keyTS[types.TimestampPrefixLen:] // remove the timestamp prefix
+		// skip latest stateInfo and latest finalized stateInfo
+		stateInfoIndex := types.StateInfoIndexFromKey(key)
+		index := skipStateInfoIndexesMap[stateInfoIndex.RollappId]
+		if index == stateInfoIndex.Index {
+			return false
+		}
+
+		store.Delete(key)
+		storeTS.Delete(keyTS)
+		return false
+	})
+}
```

Comment:
is deleting while iterating definitely cool?
I remember having some issues with this in previous sdk version

Reply:
now you're making me paranoid

Reply:
I don't see why not. What issues were you having? Were you also deleting, or modifying?

Reply:
This was a while ago when I was on cosmos hub team
I think it was sdk 45 but I can't remember tbh
I think we deleted while iterating and it caused some problem and Marko and core sdk team told us not to do it

So needs to be researched and double checked

Reply:
I believe this is the cache kv store wrapper for the iavl store, so all the writes/deletes should be executed on the actual state transition

Reply:
yeah that rings a bell actually 👍 
so probs it's fine

Reply:
nice